### ISSUE TITLE
Colors a pain to define in qbarplot

### ISSUE THREAD

**Issue by vertesy** [vertesy/ggExpress#33]
**Description**
Colors a pain to define in qbarplot.
You can't easily provide a color vector as in `wbarplot()`

**Comment by vertesy** [comment]
## Solution

qbarplot should repurpose the stupid palette and color options.

col 
1. NULL by default
2. If provided, AND == len(xxx)
`palette = rich.colors(n = length(xxx)), col = 1:length(xxx))`

**Comment by vertesy** [comment]
## Seems to need a larger revamp

```r
   col = as.character(1:3)[1],
    xlab = ...
    ...) {

# later

  df[["colour"]] <-
    if (length(col) == length(vec)) {
      as.character(col)
    } else if (hline & filtercol != 0) {
      if (filtercol == 1) (df$"value" > hline) else if (filtercol == -1) (df$"value" < hline)
    } else {
      as.character(rep(col, length(vec))[1:length(vec)])
    }
```

**Comment by vertesy** [comment]
```r
# TL;DR — ROOT CAUSE + FIX STRATEGY
#
# qbarplot() currently mixes **three incompatible color semantics**:
#  1) ggpubr::ggbarplot(palette=...)  → factor → scale_fill_manual()
#  2) explicit colors (e.g. "brown") → identity scale
#  3) numeric indices into a palette → positional lookup
#
# The bug is that `df$colour` is *sometimes numeric, sometimes logical,
# sometimes character*, and ggbarplot() always assumes **factor levels**
# that must match `names(palette)`.
#
# SOLUTION:
#   - Decide colors BEFORE ggbarplot()
#   - Convert them to **explicit hex / color strings**
#   - Use scale_fill_identity() + scale_color_identity()
#   - STOP passing `palette=` to ggbarplot() except in legacy mode
#
# This makes color behavior predictable and fixes *all* issues you described.
```

------

## Why `qbarplot()` is broken (precise diagnosis)

### 1️⃣ `ggbarplot()` expects **factor levels**, not colors

Internally `ggpubr::ggbarplot()` does this (simplified):

```r
aes(fill = colour) +
scale_fill_manual(values = palette)
```

So:

- `df$colour = "brown"`
  → ggplot looks for a palette entry named `"brown"` ❌
- `df$colour = 2`
  → treated as factor `"2"` → mapped to 2nd palette color ✔
- named palette + unnamed data
  → **no shared levels → everything gray** (exact warning you see)

This is **by design** in ggpubr, not a ggExpress bug per se.

------

### 2️⃣ `df$colour` is overloaded and inconsistent

```r
df[["colour"]] <-
  if (length(col) == length(vec)) {
    as.character(col)
  } else if (hline & filtercol != 0) {
    (df$value > hline)  # <-- LOGICAL (!!)
  } else {
    as.character(rep(col, length(vec)))
  }
```

This means `colour` can be:

| Case             | Type                    | Meaning                   |
| ---------------- | ----------------------- | ------------------------- |
| `col = 1:3`      | character `"1","2","3"` | palette indices           |
| `col = "brown"`  | character `"brown"`     | *intended* explicit color |
| `filtercol=TRUE` | logical `TRUE/FALSE`    | threshold encoding        |

**ggbarplot cannot resolve this ambiguity.**

------

## Correct mental model (what users *expect*)

You described it perfectly. Color input should be resolved like this:

1. **Explicit colors always win**

   ```r
   col = c("brown", "orange")
   ```

   → bars get those colors directly

2. **Named colors map by names**

   ```r
   c(A = "red", B = "blue")
   ```

   → matched against `names(vec)`

3. **Numeric `col` indexes into a palette**

   ```r
   col = 1:3, palette = "jco"
   ```

4. **filtercol is just a color selector**, not data

   ```r
   above = palette[1], below = palette[2]
   ```

This is **NOT how ggbarplot works**, so the fix must happen *before* calling it.

------

## The correct fix 🔧 – Resolve colors explicitly, but without handling named palettes

```r
# --- MINIMAL COLOR RESOLUTION (ggpubr-compatible) --------------------

pal <- if (length(palette_use) == 1)
  ggpubr::get_palette(palette_use, length(vec))
else palette_use

cols <- if (is.numeric(col))
  pal[rep(col, length.out = length(vec))]
else rep(col, length.out = length(vec))

keys <- as.character(seq_along(cols))
pal2 <- setNames(cols, keys)
df$colour <- keys

p <- ggpubr::ggbarplot(
  data = df, x = "names", y = "value",
  color = "colour", fill = "colour",
  palette = pal2,
  ...
)
```





------

## Result: 

| Case                        | Works?              |
| --------------------------- | ------------------- |
| `col = "brown"`             | ✅                   |
| `col = c("brown","orange")` | ✅                   |
| `col = 1:3, palette="jco"`  | ✅                   |
| named palettes              | ✅                   |
| partial name overlap        | ❌ ( and no warning) |
| filtercol logic             | ? not tested        |
|                             |                     |

------



## Working tests

````R
xxx <- 1:3
qbarplot(xxx, col = 1) # jco Blue
qbarplot(xxx, col = 2) # jco Yellow
qbarplot(xxx, palette_use = "aaas", col = 2) # AAAS Blue
qbarplot(xxx, palette_use = "aaas", col = 1:3) # AAAS Blue, red, Green

qbarplot(xxx, col = 2:1) # This works within the palette (2nd, 1st, 2nd color of )

qbarplot(xxx, col = "brown")  # brown
qbarplot(xxx, col = c("brown", "orange")) # Wtf: now it reverts back to default palette and alternating colors like `col = 2:1`


qbarplot(xxx, palette = c("brown", "orange"), col = 1) # This works as expected: brown
qbarplot(xxx, palette = c("brown", "orange")) # This does not work as expected by default because the default color argument is one, so it only uses the brown color


color_check(rich.colors(n = length(xxx)))
qbarplot(xxx, palette = rich.colors(n = length(xxx)))
qbarplot(xxx, col = rich.colors(n = length(xxx)))
qbarplot(xxx, palette = rich.colors(n = length(xxx)), col = 1:length(xxx))


names(xxx) <- letters[1:3]
(named_palette <- setNames(rich.colors(n = 6)[3:5], letters[1:3]))
color_check(named_palette)

qbarplot(xxx, palette = named_palette, col = 1:3) # As expected, all 3 cols
qbarplot(xxx, palette = named_palette, col = 3:1) # As expected, reverse  3 cols
qbarplot(xxx, palette = named_palette)  # col 1 of palette
qbarplot(xxx, col = named_palette)  # As expected, all 3 cols
qbarplot(xxx, col = rev(named_palette))  # As expected, all 3 cols
```

# Test naming overlap issues - FAIL

```r
# CASE 1 :Complete mismatch of names, no overlap between the value names and color or palette name

# "No overlap" Warning expected, Colors can be used as is.

(named_palette2 <- setNames(rich.colors(n = 6)[3:5], LETTERS[1:3])) # All caps

names(named_palette2) %in% names(xxx)

# [1] FALSE FALSE FALSE

qbarplot(xxx, col = named_palette2) # Unexpected: it uses all 3 cols as if they were not named, So no name matching is performed, and no warning is raised.
qbarplot(xxx, palette = named_palette2, col = 1:length(xxx))  # Unexpected: it uses all 3 cols as if they were not named, So no name matching is performed, and no warning is raised.
qbarplot(xxx, palette = named_palette2) # uses col1

## FAILS at name matching between value names and palette names

# CASE 2 :Partial mismatch of names, some overlap between the value names and color or palette name

# "Partial overlap" Warning expected, Non matching colors should be treated as NA, which I  believe is matched to the gray color.

(named_palette3 <- setNames(rich.colors(n = 6)[3:5], c("a", "B", "C"))) # Mixed case
names(named_palette3) %in% names(xxx)

# [1]  TRUE FALSE FALSE

qbarplot(xxx, col = named_palette3) # Unexpected: it uses all 3 cols as if they were not named, So no name matching is performed, and no warning is raised.
qbarplot(xxx, palette = named_palette3, col = 1:length(xxx))  # Unexpected: it uses all 3 cols as if they were not named, So no name matching is performed, and no warning is raised.
qbarplot(xxx, palette = named_palette3) # uses col1

````

**Comment by vertesy** [comment]
## Development and test code 

```r
# backup(qbarplot)
qbarplot <- NULL

qbarplot <- function(
    vec,
    also.pdf = FALSE, save = F, save.obj = FALSE,
    ext = MarkdownHelpers::ww.set.file.extension(default = "png", also_pdf = also.pdf),
    plot = TRUE,
    plotname = FixPlotName(substitute(vec)),
    subtitle = paste("Median:", iround(median(vec))),
    suffix = NULL,
    caption = suffix,
    filename = NULL,
    mdlink = MarkdownHelpers::unless.specified("b.mdlink", def = FALSE),
    hline = FALSE, filtercol = 1,

    palette_use = c("RdBu", "Dark2", "Set2", "jco", "npg", "aaas", "lancet", "ucscgb", "uchicago")[4],
    col = 1,

    xlab = "", xlab.angle = 45,
    logY = FALSE,
    ylim = c(0, iround(1.1 * as.numeric(max(vec, na.rm = TRUE)))),
    annotation_logticks_Y = logY,
    label = NULL,
    hide.legend = TRUE,
    legend.title = NULL,
    max.names = 100,
    limitsize = FALSE,
    grid = "y",
    ylab = NULL,
    w = qqqAxisLength(vec, factor = 0.25), h = 5,
    ...) {

  stopifnot(is.numeric(vec), length(vec) > 0L, all(is.finite(vec)))
  if (isFALSE(xlab)) xlab <- plotname
  df <- qqqNamed.Vec.2.Tbl(namedVec = vec, strip.too.many.names = FALSE)
  if (length(unique(df$"names")) == 1) df$"names" <- as.character(1:length(vec))


  # --- Name-aware color handling (minimal, ggpubr-native) ----------------
  # decide which vector carries names for color semantics

  col_src <- if (!is.null(names(col))) col else palette_use
  if (!is.null(names(col_src)) && !is.null(names(vec))) {
    name_overlap <- intersect(names(col_src), names(vec))

    Stringendo::warnifnot(
      "No overlap between names(vec) and names(col) / names(palette)! Names are ignored, match by order!" = length(name_overlap) > 0,
      "Partial overlap between names(vec) and names(col). Non-matching names are grey (NA)." = (length(name_overlap) == 0 || length(name_overlap) == length(vec))
    )

    if (length(name_overlap) == length(vec)) { # IMPORTANT: only mutate col on FULL or ZERO overlap
      col <- col_src[names(vec)]
    } else if (length(name_overlap) == 0) {
      col <- unname(col)
    }
    # partial overlap: warn only, do NOT touch col
  }


  # Handling colors ________________________________________________________________

  # Palette argument
  pal <- if (length(palette_use) == 1)
    ggpubr::get_palette(palette_use, length(vec)) # For a name of a palette or a single color.
  else
    palette_use

  # Color argument
  cols <- if (is.numeric(col))
    pal[rep(col, length.out = length(vec))]
  else rep(col, length.out = length(vec))

  keys <- as.character(seq_along(cols))
  pal2 <- setNames(cols, keys)
  df$colour <- keys

  p <- ggpubr::ggbarplot(
    data = df, x = "names", y = "value",
    title = plotname, xlab = xlab,
    subtitle = subtitle,
    caption = caption,
    color = "colour", fill = "colour",
    palette = pal2,
    label = label,
    ylim = ylim,
    ...
  ) +
    ggplot2::labs(y = ylab) +
    ggpubr::grids(axis = "y") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = xlab.angle, hjust = 1))


  if (grid %in% c("xy", "x", "y")) p <- p + ggpubr::grids(axis = grid)

  if (length(vec) > max.names) p <- p + ggplot2::guides(x = "none")
  if (hide.legend) p <- p + ggplot2::theme(legend.position = "none")
  if (!is.null(legend.title)) p <- p + ggplot2::guides(fill = ggplot2::guide_legend(title = legend.title), color = "none") # Hide the color legend

  if (!isFALSE(hline)) p <- p + ggplot2::geom_hline(yintercept = hline)
  if (logY) p <- p + ggplot2::scale_y_log10()
  if (annotation_logticks_Y) p <- p + ggplot2::annotation_logticks(sides = "l")

  file_name <- if (!is.null(filename)) {
    filename
  } else {
    FixPlotName(plotname, suffix, flag.nameiftrue(logY), "bar", ext)
  }

  if (save) {
    qqSave(
      ggobj = p, title = plotname, fname = file_name, ext = ext,
      w = w, h = h, limitsize = limitsize, also.pdf = also.pdf, save.obj = save.obj
    )
  }
  if (mdlink & save) qMarkdownImageLink(file_name)
  if (plot) print(p)
}





xxx <- 1:3
qbarplot(xxx, col = 1) # jco Blue
qbarplot(xxx, col = 2) # jco Yellow
qbarplot(xxx, palette_use = "aaas", col = 2) # AAAS Blue
qbarplot(xxx, palette_use = "aaas", col = 1:3) # AAAS Blue, red, Green

qbarplot(xxx, col = 2:1) # This works within the palette (2nd, 1st, 2nd color of )

qbarplot(xxx, col = "brown")  # brown
qbarplot(xxx, col = c("brown", "orange")) # Wtf: now it reverts back to default palette and alternating colors like `col = 2:1`


qbarplot(xxx, palette = c("brown", "orange"), col = 1) # This works as expected: brown
qbarplot(xxx, palette = c("brown", "orange")) # This does not work as expected by default because the default color argument is one, so it only uses the brown color


color_check(rich.colors(n = length(xxx)))
qbarplot(xxx, palette = rich.colors(n = length(xxx)))
qbarplot(xxx, col = rich.colors(n = length(xxx)))
qbarplot(xxx, palette = rich.colors(n = length(xxx)), col = 1:length(xxx))


# Test using perfectly matching naming
names(xxx) <- rev(letters[1:3])
(named_palette <- setNames(rich.colors(n = 6)[3:5], letters[1:3]))
color_check(named_palette)

qbarplot(xxx, palette = named_palette, col = 1:3) # As expected, all 3 cols
qbarplot(xxx, palette = named_palette, col = 3:1) # As expected, reverse  3 cols
qbarplot(xxx, palette = named_palette)  # col 1 of palette

qbarplot(xxx, col = named_palette)  # As expected, all 3 cols, And in reverse order of the name palette as expected by the reverse matching name
qbarplot(xxx, col = rev(named_palette))  # As expected, this does not change anything, because the matching is by name, not by order


# Test naming overlap issues - FAIL

# CASE 1 :Complete mismatch of names, no overlap between the value names and color or palette name
# "No overlap" Warning expected, Colors can be used as is.

(named_palette2 <- setNames(rich.colors(n = 6)[3:5], LETTERS[1:3])) # All caps

names(named_palette2) %in% names(xxx)
# [1] FALSE FALSE FALSE

qbarplot(xxx, col = named_palette2)
qbarplot(xxx, palette = named_palette2, col = 1:length(xxx))  # Unexpected: it uses all 3 cols as if they were not named, So no name matching is performed, and no warning is raised.
qbarplot(xxx, palette = named_palette2) # uses col1

## FAILS at name matching between value names and palette names

# CASE 2 :Partial mismatch of names, some overlap between the value names and color or palette name
# "Partial overlap" Warning expected, Non matching colors should be treated as NA, which I  believe is matched to the gray color.

(named_palette3 <- setNames(rich.colors(n = 6)[3:5], c("a", "B", "C"))) # Mixed case
names(named_palette3) %in% names(xxx)
# [1]  TRUE FALSE FALSE

qbarplot(xxx, col = named_palette3)
qbarplot(xxx, palette = named_palette3, col = 1:3)  # Unexpected: it uses all 3 cols as if they were not named, So no name matching is performed, and no warning is raised.
qbarplot(xxx, palette = named_palette3) # uses col1

```
